cpp/include/cudf/strings/detail/modify_strings.cuh

```diff
@@ -0,0 +1,110 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <strings/utilities.cuh>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+/**
+ * @brief Generic string modification in two passes: 1st pass probes for memory load requirements;
+ * 2nd pass executes string modification.
+ *
+ * @tparam device_probe_functor Functor for probing memory requitements;
+ * must implement `__device__ int32_t operator()(size_type idx) const`
+ * @tparam device_execute_functor Functor for executing string modification; must
+ * implement `__device__ int32_t operator()(size_type idx)`
+ * @tparam ...Types Types of possible additional arguments to be forwarded
+ * to the probe / execute functors (pre-condition: must both take the same trailling pack of
+ * arguments, in addition to their required args)
+ *
+ * @param strings Number Column of strings to apply the modifications on;
+ * it is not modified in place; rather a new column is returned instead
+ * @param mr Memory resource to use
+ * (cannot be a default argument because of the variadic pack);
+ * @param stream Stream to use for any kernel calls.
+ * (cannot be a default argument because of the variadic pack);
+ * @param ...args Additional arguments to be forwarded to
+ * the probe / execute constructors (can be empty);
+ * @return modified strings column
+ */
+template <typename device_probe_functor, typename device_execute_functor, typename... Types>
+std::unique_ptr<column> modify_strings(strings_column_view const& strings,
+                                       rmm::mr::device_memory_resource* mr,
+                                       cudaStream_t stream,
+                                       Types&&... args)
+{
+  auto strings_count = strings.size();
+  if (strings_count == 0) return detail::make_empty_strings_column(mr, stream);
+
+  auto execpol = rmm::exec_policy(stream);
+
+  auto strings_column  = column_device_view::create(strings.parent(), stream);
+  auto d_column        = *strings_column;
+  size_type null_count = strings.null_count();
+
+  // copy null mask
+  rmm::device_buffer null_mask = copy_bitmask(strings.parent(), stream, mr);
+  // get the lookup tables used for case conversion
+
+  device_probe_functor d_probe_fctr{d_column, std::forward<Types>(args)...};
```

Comment:
Instead of passing a parameter pack to `modify_strings`, why not require passing in pre-constructed functor objects?

```
template <typename device_probe_functor, typename device_execute_functor>
std::unique_ptr<column> modify_strings(strings_column_view const& strings,
                                       rmm::mr::device_memory_resource* mr,
                                       cudaStream_t stream,
                                       device_probe_functor P, device_execute_functor E)

Reply:
Because these functors cannot be _pre_-constructed. They require data that is not yet available at the time of `modify_strings()` call (buffers, etc. that are constructed _inside_ `modify_strings()`). The idea is to avoid setters, which would break RAII.

That being said, I had a conversation with @davidwendt on this topic. And he proposed a better solution, which we agreed I would deliver in a separate PR after this gets merged (to address more pressing goals, first).